ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java

```diff
@@ -281,4 +262,121 @@ private void unloadErrorsFromCache(final ImmutableMap<K, ComposableFuture<V>> in
+    private final int ttl;
+    private final TimeUnit unit;
+    private final MetricFactory metricFactory;
+    private final String cacheName;
+    private CacheLoader<K, V> loader;
+    private boolean failOnMissingEntries = false;
+    private long refreshAfterWriteDuration;
+    private TimeUnit refreshAfterWriteUnit;
+    private long loadTimeout = DEFAULT_LOAD_TIMEOUT;
+    private TimeUnit loadTimeoutUnit = DEFAULT_LOAD_TIMEOUT_UNIT;
+    private int refreshThreadCount = DEFAULT_REFRESH_THREAD_COUNT;
+
+    public Builder(final int maximumSize, final int ttl, final TimeUnit unit, final MetricFactory metricFactory, final String cacheName) {
+      this.maximumSize = maximumSize;
+      this.ttl = ttl;
+      this.unit = unit;
+      this.metricFactory = metricFactory;
+      this.cacheName = cacheName;
+    }
+
+    public Builder<K, V> withLoader(final CacheLoader<K, V> loader) {
+      this.loader = loader;
+      return this;
+    }
+
+    public Builder<K, V> withLoadTimeout(final long loadTimeout, final TimeUnit loadTimeoutUnit) {
+      this.loadTimeout = loadTimeout;
+      this.loadTimeoutUnit = loadTimeoutUnit;
+      return this;
+    }
+
+    public Builder<K, V> failOnMissingEntries() {
+      this.failOnMissingEntries = true;
+      return this;
+    }
+
+    public Builder<K, V> refreshAfterWrite(long duration, TimeUnit unit) {
+      this.refreshAfterWriteDuration = duration;
+      this.refreshAfterWriteUnit = unit;
+      return this;
+    }
+
+    public Builder<K, V> withRefreshThreadCount(int refreshThreadCount) {
+      this.refreshThreadCount = refreshThreadCount;
+      return this;
+    }
+
+    public LocalAsyncCache<K, V> build() {
+      if (loader == null) {
+        return new LocalAsyncCache<>(maximumSize, ttl, unit, metricFactory, cacheName);
+      }
+      return new LocalAsyncCache<>(maximumSize, ttl, unit, loader, metricFactory, cacheName, failOnMissingEntries,
+              loadTimeout, loadTimeoutUnit, refreshAfterWriteDuration, refreshAfterWriteUnit, refreshThreadCount);
+    }
+  }
+
+  private class InternalCacheLoader extends com.google.common.cache.CacheLoader<K, ComposableFuture<V>> {
+
+    private final CacheLoader<K, V> loader;
+    private final String cacheName;
+    private final boolean failOnMissingEntries;
+    private final long loadTimeout;
+    private final TimeUnit loadTimeoutUnit;
+
+    InternalCacheLoader(final CacheLoader<K, V> loader, final String cacheName, final boolean failOnMissingEntries,
+                        final long loadTimeout, final TimeUnit loadTimeoutUnit) {
+      this.loader = loader;
+      this.cacheName = cacheName;
+      this.failOnMissingEntries = failOnMissingEntries;
+      this.loadTimeout = loadTimeout;
+      this.loadTimeoutUnit = loadTimeoutUnit;
+    }
+
+    @Override
+    public ComposableFuture<V> load(@Nonnull final K key) {
+      return loader.load(cacheName, key).withTimeout(loadTimeout, loadTimeoutUnit).materialize();
```

Comment:
Why did you add withTimeout?

Reply:
I added timeout because LoadingCacheDelegate has the option to set timeout to the load and i felt that it was missing here.
Don't you agree?